network: bridge: add support for configuring locked ports

KanjiMonster opened the pull request:
"Recently" (as of 5.18) the Linux kernel gained the ability of locking bridge ports to restrict network access to authenticated hosts only.

This is implemented by disabling automated learning and dropping incoming traffic from unknown hosts. User space is then expected to add fdb entries for authenticated hosts. Once a fdb entry exist, traffic for that host will be forwarded as expected.

This was later extended with "Mac Authentication Bypass", where the locking was extended to fdb entries. In this mode the kernel adds fdb entries again automatically, but they are locked by default.

To properly configure this, add two network options and one netdev option:

* `NoLinkLocalLearning=` to prevent the kernel from creating unlocked entries based on link-local traffic, which would bypass any authentication. Needed when enabling learning on a locked port.
* `Locked=` to allow setting a bridge port to locked.
* `MAB=` to allow enabling Mac Authentication Bypass on a port. Requires learning to be enabled on the port as well (and consequently `NoLinkLocalLearning` enabled on the bridge).

An authenticator (e.g. hostapd) is still needed to do the actual authentication, the kernel only provides the access control.

KanjiMonster commented:
Added missing initialization of `Network.mab` to -1.

yuwata commented on src/network/netdev/bridge.c:
```diff
@@ -142,6 +143,18 @@ static int netdev_bridge_post_create_message(NetDev *netdev, sd_netlink_message
 
+        if (b->no_ll_learn >= 0) {
+                bm.optmask |= 1 << BR_BOOLOPT_NO_LL_LEARN;
+                if (b->no_ll_learn)
+                        bm.optval |= 1 << BR_BOOLOPT_NO_LL_LEARN;
```
Please use `SET_FLAG()` macro:
```c
SET_FLAG(bm.optval, 1 << BR_BOOLOPT_NO_LL_LEARN, b->no_ll_learn);
```

yuwata commented on man/systemd.netdev.xml:
```diff
@@ -456,6 +456,16 @@
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>NoLinkLocalLearning=</varname></term>
```
Please make the setting positive, i.e. `LinkLocalLearning=`.
We typically do not use negative words in setting.

yuwata commented on src/network/netdev/bridge.h:
```diff
@@ -21,6 +21,7 @@ typedef struct Bridge {
         uint8_t igmp_version;
         uint32_t fdb_max_learned;
         bool fdb_max_learned_set;
+        int no_ll_learn;
```
Then, also invert the variable name.

yuwata commented on man/systemd.network.xml:
```diff
@@ -4624,6 +4624,15 @@ ServerAddress=192.168.0.1/24</programlisting>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>Locked=</varname></term>
```
commit message: Please use the full hash of the kernel commit, and also mention that is in v5.18 .

yuwata commented on src/network/networkd-network.h:
```diff
@@ -297,6 +297,7 @@ struct Network {
         uint32_t cost;
         uint16_t priority;
         MulticastRouter multicast_router;
+        int locked;
```
For a new element, please prefix with `bridge_`, like other elements e.g. `dhcp_xyz`.
Yes, most of the above elements are not prefixed, but for new one, please.

yuwata commented on man/systemd.network.xml:
```diff
@@ -4630,6 +4630,15 @@ ServerAddress=192.168.0.1/24</programlisting>
 
+          <xi:include href="version-info.xml" xpointer="v258"/>
+        </listitem>
+        <term><varname>MAB=</varname></term>
```
Please do not use abbreviate word, so `MACAuthenticationBypass=`.

commit message: please use the full hash and mention that it is in v6.2 .

yuwata commented on src/network/networkd-network.h:
```diff
@@ -298,6 +298,7 @@ struct Network {
         uint16_t priority;
         MulticastRouter multicast_router;
         int locked;
+        int mab;
```
Also here. `mab` -> `bridge_mab`

yuwata reviewed (changes requested):
We usually do not use Signed-off-by tag. Unless your company or so explicitly request the tag, please drop it in each commit.

Please add test cases for the new features in test/test-network/systemd-networkd-tests.py . Maybe, test_bridge() for NO_LL_LEARN, and test_bridge_property() for locked and mab features, are the best place. But you can add any new ones.

Thank you!

KanjiMonster commented:
> We usually do not use Signed-off-by tag. Unless your company or so explicitly request the tag, please drop it in each commit.

Ah sorry, it's mostly a habit from (mainly) doing kernel and OE stuff.

> 
> Please add test cases for the new features in test/test-network/systemd-networkd-tests.py . Maybe, test_bridge() for NO_LL_LEARN, and test_bridge_property() for locked and mab features, are the best place. But you can add any new ones.

I can do that, but how to I test it? I Can't find anything about how to run the tests in https://github.com/systemd/systemd/blob/main/docs/HACKING.md or https://github.com/systemd/systemd/blob/main/docs/CONTRIBUTING.md (I had hoped that the instruction to run tests would also tell me how).

yuwata commented:
You can see the brief instruction in the header of systemd-networkd-test.py.

KanjiMonster commented:
Updated, hopefully adressing all review comments:
- renamed `NoLinkLocalLearning` to `LinkLocalLearning` (and consequently `no_ll_learn` to `linklocal_learn`)
- use `SET_FLAG()`
- renamed `MAB` to `MACAuthenticationBypass` (and consequently `mab` to `mac_authentication_bypass`)
- added a `bridge_` prefix to `locked` and `mac_authentication_bypass`
- dropped my SOBs
- Used full kernel commit hashes and kernel versions
- Extended `systemd-networkd-tests.py` tests for the new options
- Rebased on latest `main`

KanjiMonster commented:
Pushed a fix for the noble failure. The iproute2 there is too old and doesn't know 'mab' yet, and the option isn't exposed via sysfs.

yuwata reviewed (approved):
LGTM. Thank you!
